### PR TITLE
[SPARK-34512][BUILD][SQL] Upgrade built-in Hive to 2.3.9

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr upgrades built-in Hive to 2.3.9. Hive 2.3.9 changes:
- [HIVE-17155] - findConfFile() in HiveConf.java has some issues with the conf path
- [HIVE-24797] - Disable validate default values when parsing Avro schemas 
- [HIVE-24608] - Switch back to get_table in HMS client for Hive 2.3.x
- [HIVE-21200] - Vectorization: date column throwing java.lang.UnsupportedOperationException for parquet
- [HIVE-21563] - Improve Table#getEmptyTable performance by disabling registerAllFunctionsOnce
- [HIVE-19228] - Remove commons-httpclient 3.x usage

### Why are the changes needed?

Fix regression caused by AVRO-2035.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @sunchao

File: pom.xml [hunk 1]
```diff
@@ -300,6 +300,17 @@
         <enabled>false</enabled>
       </snapshots>
     </repository>
+    <repository>
+      <id>staged</id>
+      <name>hive-staged-releases</name>
+      <url>https://repository.apache.org/content/repositories/staging/</url>
```
Comment (srowen) [inline pom.xml:306]: Is 2.3.9 not released yet?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 2]
```diff
@@ -669,7 +669,7 @@ private[sql] class TestHiveSessionStateBuilder(
 }
 
 private[hive] object HiveTestJars {
-  private val repository = SQLConf.ADDITIONAL_REMOTE_REPOSITORIES.defaultValueString.split(",")(0)
+  private val repository = SQLConf.ADDITIONAL_REMOTE_REPOSITORIES.defaultValueString.split(",").last
```
Comment (srowen) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:672]: Out of curiosity why is this needed - the staging repo?

File: pom.xml (hunk 1 above)
Comment (wangyum) [reply pom.xml:306]: Yes.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala (hunk 2 above)
Comment (wangyum) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:672]: Need download jars from `https://repository.apache.org/content/repositories/staging/`.

Comment (srowen) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:672]: OK so when it's released we can remove this? seems fine yes.

**Comment by dongjoon-hyun** [comment]
Thank you, @wangyum . It seems that the downloading still have some issue.
```
sbt.ForkMain$ForkError: java.lang.RuntimeException: [unresolved dependency: org.apache.hive.hcatalog#hive-hcatalog-core;2.3.9: not found]
	at org.apache.spark.deploy.SparkSubmitUtils$.resolveMavenCoordinates(SparkSubmit.scala:1446)
```

**Comment by dongjoon-hyun** [comment]
Hi, @wangyum . 

@sunchao uploaded Apache Hive 2.3.9 to the Maven Central. Could you revise this PR?
- https://repo1.maven.org/maven2/org/apache/hive/hive-exec/2.3.9/

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs)
Thank you for updating.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.2.0.
Thank you so much, @wangyum and @sunchao !
